Tasks/Gulp/gulptask.ts

```diff
@@ -2,118 +2,152 @@
+			publishTestResults(publishJUnitResults, testResultsFiles);
 			tl.debug('taskRunner fail');
-			tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+			throw new Error(tl.loc('GulpFailed', err.message));
 		})
-	} else {
-		tl.setResult(tl.TaskResult.Succeeded, tl.loc('GulpReturnCode', code));
 	}
-}).fail(function (err) {
-	publishTestResults(publishJUnitResults, testResultsFiles);
-	tl.debug('taskRunner fail');
-	tl.setResult(tl.TaskResult.Failed, tl.loc('GulpFailed', err.message));
-})
+}
 
 function publishTestResults(publishJUnitResults, testResultsFiles: string) {
-    if (publishJUnitResults) {
-        //check for pattern in testResultsFiles
-        if (testResultsFiles.indexOf('*') >= 0 || testResultsFiles.indexOf('?') >= 0) {
-            tl.debug('Pattern found in testResultsFiles parameter');
-            var buildFolder = tl.getVariable('System.DefaultWorkingDirectory');
-            var allFiles = tl.find(buildFolder);
-            var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, { matchBase: true });
-        }
-        else {
-            tl.debug('No pattern found in testResultsFiles parameter');
-            var matchingTestResultsFiles = [testResultsFiles];
-        }
-        if (!matchingTestResultsFiles || matchingTestResultsFiles.length == 0) {
-            tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
-            return 0;
-        }
-        var tp = new tl.TestPublisher("JUnit");
+	if (publishJUnitResults) {
+		//check for pattern in testResultsFiles
+		if (/[\*?+@!{}\[\]]/.test(testResultsFiles)) {
+			tl.debug('Pattern found in testResultsFiles parameter');
+			var matchingTestResultsFiles = tl.findMatch(tl.getVariable('System.DefaultWorkingDirectory'), testResultsFiles, null, { matchBase: true });
+		}
+		else {
+			tl.debug('No pattern found in testResultsFiles parameter');
+			var matchingTestResultsFiles = [testResultsFiles];
+		}
+		if (!matchingTestResultsFiles || matchingTestResultsFiles.length == 0) {
+			tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
+			return 0;
+		}
+		var tp = new tl.TestPublisher("JUnit");
 		try {
 			tp.publish(matchingTestResultsFiles, true, "", "", "", true);
 		} catch (error) {
 			tl.warning(error);
 		}
-    }
+	}
 }
 
-function publishCodeCoverage(summaryFile) {
+function publishCodeCoverage(summaryFile, reportDirectory) {
 	try {
 		var ccPublisher = new tl.CodeCoveragePublisher();
 		ccPublisher.publish('cobertura', summaryFile, reportDirectory, "");
 	} catch (error) {
 		tl.warning(error);
 		throw error;
 	}
-}
\ No newline at end of file
+}
+
+function getGulpFiles(gulpFilePath: string): string[] {
+	var gulpFiles = [];
+
+	if (/[\*?+@!{}\[\]]/.test(gulpFilePath)) {
+		gulpFiles = tl.findMatch(tl.getVariable("System.DefaultWorkingDirectory"), gulpFilePath);
```

Comment:
I am concerned here about the input type. `gulpFilePath` is a single-line input.

What is the common scenario for multi-line input with this task? If it's `**/gulpfile.js`, then I'm worried whether that will accidentally resolve files under node_modules. It's difficult in a single-line input to filter results from within node_modules. It's easy in a multi-line input.

What do you think about that scenario? Are my concerns legitimate?